### PR TITLE
Added shutdown crate

### PR DESCRIPTION
**Pull request by sdbondi**
## Description

The Shutdown struct from tari_comms will be more generally useful to
shutdown the various long running tasks. These include:

- Comms
- DHT
- Domain services
- Wallet application

The shutdown struct could be described as a oneshot broadcast. This version has been greatly simplified from the tari_comms one.  Once triggered, all listeners to the signal will trigger.

Potential alternative locations: tari_utilities - only downside is that futures and tokio get added as deps

## Motivation and Context

Related to #869 

## How Has This Been Tested?

Unit tests 

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: infrastructure/shutdown/src/lib.rs [hunk 1]
```diff
@@ -20,90 +20,48 @@
...
-            }
-
-            for result in join_all(receivers).timeout(self.timeout).await.map_err(|_| ())? {
-                result.map_err(|_| ())?
-            }
+    pub fn trigger(&mut self) -> Result<(), ()> {
+        if self.trigger.is_some() {
+            self.trigger
+                .take()
+                .expect("previously checked trigger is Some")
```
Comment (CjS77) [inline infrastructure/shutdown/src/lib.rs]: Could also use `unreachable!` macro

File: infrastructure/shutdown/src/lib.rs [hunk 2]
```diff
@@ -114,21 +72,24 @@ impl Shutdown {
     }
 
     pub fn is_triggered(&self) -> bool {
-        self.is_triggered
+        self.trigger.is_none()
+    }
+}
+
+impl Drop for Shutdown {
```
Comment (CjS77) [inline infrastructure/shutdown/src/lib.rs]: Sjoe. You should document that the trigger gets fired when `Self` goes out of scope. This isn't super intuitive behavior without diving into the code.

File: infrastructure/shutdown/src/lib.rs [hunk 3]
```diff
@@ -137,25 +98,41 @@ mod test {
         let signal = shutdown.new_signal();
         assert_eq!(shutdown.is_triggered(), false);
         rt.spawn(async move {
-            signal.await.unwrap().signal().unwrap();
+            signal.await.unwrap();
         });
-        rt.block_on(shutdown.trigger()).unwrap();
+        shutdown.trigger().unwrap();
+        // Trigger > once is a noop
```
Comment (CjS77) [inline infrastructure/shutdown/src/lib.rs]:
```suggestion
        // Shutdown::trigger() is idempotent
```

File: infrastructure/shutdown/src/lib.rs (hunk 1 above)
Comment (sdbondi) [reply infrastructure/shutdown/src/lib.rs]: Much cleaner thanks!

File: infrastructure/shutdown/src/lib.rs (hunk 2 above)
Comment (sdbondi) [reply infrastructure/shutdown/src/lib.rs]: You're right! Doing it this way made sense to me when writing tests and using it in applications, but could definitely cause confusion if one is not aware of it

**Pull request merged**